manager/src/main/java/io/gingersnapproject/metrics/micrometer/CacheManagerMicrometerMetrics.java

```diff
@@ -99,7 +99,33 @@ public void localHit(boolean localHit) {
       }
 
       private void recordLatency(TimerMetric metric) {
-         timers.get(metric).record(System.nanoTime() - startTimeNanos, TimeUnit.NANOSECONDS);
+         Objects.requireNonNull(perRulesMetrics.get(rule), String.format("Rule '%s' not registered", rule))
+               .timers.get(metric).record(System.nanoTime() - startTimeNanos, TimeUnit.NANOSECONDS);
+      }
+   }
+
+   private class RuleMetrics {
+      private final String rule;
+      private final EnumMap<TimerMetric, Timer> timers = new EnumMap<>(TimerMetric.class);
+      private final List<Meter.Id> meters = new ArrayList<>();
+
+      private RuleMetrics(String rule) {
+         this.rule = rule;
+      }
+
+      public void register(LoadingCache<?, ?> cache) {
+         var metrics = Arrays.stream(PerRuleGaugeMetric.values())
+               .map(gauge -> gauge.registerRule(registry, rule, cache))
+               .toList();
+
+         meters.addAll(metrics);
+         for (TimerMetric metric : TimerMetric.values()) {
+            timers.put(metric, createTimer(registry, metric.metricName(rule), metric.description()));
```

Comment:
you need to collect the `Id`s so you can remove them.

Reply:
I added an iteration in the `RuleMetrics#unregister` method.